### ISSUE TITLE
Attempting to proxy localhost when proxy is selected "for all URLs"

### ISSUE THREAD

**Issue by loquinator3000** [foxyproxy/firefox-extension#159]
When a proxy is selected "for all URLs," connections to localhost are proxied as well, usually resulting in a failed connection because localhost is usually not routable. This makes, for example, any web consoles for local applications inaccessible so long as there is a proxy enabled "for all URLs". I suggest adding a systemwide or per-proxy option to never attempt proxying localhost, even when there is a proxy selected "for all URLs".

**Comment by ericjung** [comment]
Hi,

On the Patterns page, please click the "Add" button next to "Add black patterns to prevent this proxy being used for localhost & intranet/private IP addresses". Here are two screenshots:

![image](https://user-images.githubusercontent.com/5875610/134773541-61875959-b23b-4bff-ae9a-23f2fe26eef5.png)

After clicking, 3 patterns are added. This included a pattern for localhost:

![image](https://user-images.githubusercontent.com/5875610/134773546-f92517f9-d089-404f-9172-8fcf77b6358f.png)

**Comment by loquinator3000** [comment]
That's the problem, FoxyProxy ignores everything on that page, including these blacklist rules, when a proxy is set to be used "for all URLs" like so:
![image](https://user-images.githubusercontent.com/70460004/134784142-7f481502-737b-4c80-8cda-9068b19e5b6e.png)
![image](https://user-images.githubusercontent.com/70460004/134784400-e027fcb8-e171-4ee4-ba92-41692fbfde7b.png)
I do have these blacklist rules set anyway.

**Comment by ericjung** [comment]
I see. I missed the "For All URLs" part, sorry. I realize this is not the most intuitive, but:

1. Set FoxyProxy to "Use Enabled Proxies by Patterns and Order"
2. Create a new proxy settings ("Add") button on main page.
3. Set Proxy Type to Direct (no proxy)
![image](https://user-images.githubusercontent.com/5875610/134787375-a6d22566-7d06-4b3f-8b08-12728b07acf3.png)

4. create whitelist pattersn for localhost, 127.0.0.1, internal IP addresses, etc:

(I copied these patterns from the "Add" button next to "Add black patterns to prevent this proxy being used for localhost & intranet/private IP addresses" then deleted those blacklist patterns:

![image](https://user-images.githubusercontent.com/5875610/134787401-0b958081-2917-4c2d-88ac-b5638544fe2c.png)

I realize that is cumbersome. Would it be easier if there was an equivalent "add" button in the whitelist patterns section? I'm not really willing to do global patterns; that's been asked about before. It causes way too much confusion for people who use patterns. "Well, this pattern applies but NOT if you first have a global pattern that collides with it."

**Comment by loquinator3000** [comment]
Unfortunately, adding this "direct proxy" doesn't solve the problem, since it happens when a proxy is set "for all URLs" which doesn't consider other proxies or any patterns.

I understand your aversion to global patterns, however the fact that there is a button to quickly add localhost and private IP addresses to a proxy's blacklist is already proof that it is very common for proxies to be switched off for these addresses. I could suggest just adding a global checkbox somewhere outside of any particular proxy. This checkbox could be labelled "never attempt to proxy local IPs", and in addition to avoiding having to add these very common rules to each new proxy, it would also prevent all proxies from proxying local IPs when they are selected in the FoxyProxy button menu. I imagine such a button would match the same URLs as the auto-added black patterns.

However, I see a problem with that proposal: this extension does not have a global settings page, where this global checkbox could go. I assume this is intentional and due to a philosophy of making proxies highly configurable by themselves. Going by this philosophy, I instead suggest a checkbox similar to the one above, but in each proxy's patterns page. This checkbox, when selected, would to stop only that specific proxy on a localhost or private IP address, even when it is selected in the extension's button menu. This checkbox can either replace or exist alongside the button to add black patterns for local IPs. that's up to you. This way, each proxy could individually choose if it wants to not proxy local IPs when it is selected from the extension's button menu, without affecting any others that might want to work even on local IPs etc always. When this checkbox is enabled, I imagine the text next to that proxy would change from "for all URLs" to "for all URLs except local IPs" like this:
![Screenshot_20210928_230743](https://user-images.githubusercontent.com/70460004/135201789-a267dd93-7cfa-4b1b-9254-7b529c51d412.png)

Ultimately, I use this extension for the fast proxy switching in the FoxyProxy button as much as the highly customizable patterns specific to each proxy. I understand your view that global patterns cause confusion, however I was similarly confused when I found that switching to a specific proxy would cause my local web consoles to break, even when the appropriate entries were in the blacklist. "Well, this blacklist pattern applies, but not if the proxy is selected in the proxy switcher". I realize the value in being able to temporarily switch on a proxy for all URLs when you might otherwise want them to work for specific ones. In fact, I personally have *.onion and *.i2p patterns my respective proxies' whitelists, and can switch to Tor or I2P proxy when I wanna be extra anonymous on any URL. I just feel local IPs should be able to be an exception to "for all URLs". I hope I found a reasonable middle ground between what I want to do with your extension and how you want your extension to be.

tldr: maybe add checkbox for each proxy to stop on local IPs, even when proxy is selected in switcher

**Comment by loquinator3000** [comment]
@ericjung Waiting for your response

**Comment by ericjung** [comment]
> I instead suggest a checkbox similar to the one above, but in each proxy's patterns page. This checkbox, when selected, would stop only that specific proxy on a localhost or private IP address, even when it is selected in the extension's button menu. 

Ok, I like this. It would have to be yet another checkbox on the patterns page because some people may want to proxy localhost and intranet IP addresses through a proxy server (obviously running inside the same network).

The "(for all URLs except local IPs)" text is key.

**Comment by loquinator3000** [comment]
Maybe a better and shorter phrase would be "(for all URLs except on LAN)" since that would imply localhost (which is not an ip address) would also be included.

**Comment by erosman** [comment]
_localhost_ should not be proxied as rule.

![connection-settings](https://user-images.githubusercontent.com/6458876/136683479-68622d0b-e557-4889-b96c-ac358cd3ff4f.jpg)

**Comment by loquinator3000** [comment]
It is when youre using this extension... that's from firefox's internal proxy config, which has no influence over this extension...

**Comment by erosman** [comment]
In case of _localhost, 127.0.01/8_, I can think of no valid reason for a connection to go according to the following:

_browser -> 3rd party proxy server over the internet -> back to user's own computer on localhost_

In case of _LAN_ limited cases might be argued e.g. testing by intranet administrator, however other cases could appear as remote hacking the _LAN_ and a security risk. By default, _LAN_ shouldn't be proxied.

**Comment by loquinator3000** [comment]
I agree. Maybe this checkbox i proposed should be enabled by default.

**Comment by ericjung** [comment]
> browser -> 3rd party proxy server over the internet -> back to user's own computer on localhost

Think of this situation:

browser -> proxy server running on localhost -> website running on localhost

Use case: developer sniffing traffic to/from code he either wrote or has the source code for

**Comment by erosman** [comment]
> Use case: developer sniffing traffic to/from code he either wrote or has the source code for

That is an extremely limited use case and in that case, there are other options available, especially to a developer. ;)

**Comment by erosman** [comment]
> I suggest adding a systemwide or per-proxy option to never attempt proxying localhost, even when there is a proxy selected "for all URLs".

Global Exclude feature is available in [FoxyProxy v8.0](https://github.com/foxyproxy/browser-extension) (once released).

**Comment by erosman** [comment]
For the time being, a default bypass/past-through for `localhost` is added to the FoxyProxy v8.0.

See also:
[connections to localhost are forwarded through the proxy](https://github.com/foxyproxy/browser-extension/issues/20)

**Comment by vortigont** [comment]
How can I disable that "default bypass/past-through for localhost"?
I NEED to proxy `localhost` but it is silently bypassed, adding pattern "*" does not helps here :(

**Comment by erosman** [comment]
@vortigont Please note [RFE: Add an option to disable "localhost" bypass](https://github.com/foxyproxy/browser-extension/issues/178#issuecomment-2642535132)
